### ISSUE TITLE
RFC: Internet Explorer 11 support deprecation and removal

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#41840]
# TL;DR

The Angular team is deprecating support for Internet Explorer 11 in Angular v12 (to be released in May 2021 and supported through November 2022), and plans to remove support for this browser in Angular v13 (late 2021).

# Why?

The worldwide usage of Internet Explorer 11 (IE11) has been steadily declining for years, and is currently at [0.73% according to StatCounter](https://gs.statcounter.com/browser-market-share#monthly-202103-202103-bar) and [1.70% according to NetMarketShare](https://netmarketshare.com/browser-market-share.aspx?options=%7B%22filter%22%3A%7B%7D%2C%22dateLabel%22%3A%22Custom%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22browserVersion%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22browsersDesktop%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222021-03%22%2C%22dateEnd%22%3A%222021-03%22%2C%22segments%22%3A%22-1000%22%7D), meaning that a very small fraction of the overall world population uses this browser and benefits from our effort to support it.

At the same time, IE11 is the last non-evergreen browser that Angular [currently supports](http://angular.io/guide/browser-support), and due to the feature gap between IE11 and evergreen browsers, it is becoming increasingly difficult to provide modern features in Angular while also preserving IE11 compatibility.

Since some of the core functionality of the framework, Material components, and CLI must work on IE11, supporting this legacy browser has resulted in complexity, compromises, or even in our inability to [offer solutions that would take advantage of modern web APIs](https://github.com/angular/components/issues/7374).

# Community feedback

We'd like the community to weigh in on the timeline of removal of IE11 support.

Angular v12 is planned to be the last version of Angular that supports IE11, this version will be:

- released in May 2021
- actively supported through November 2021, and
- supported under LTS through November 2022

See Angular's [support policy & schedule](https://angular.io/guide/releases#support-policy-and-schedule) for more information about differences between active and LTS support, and other support policy info.

This timeline seeks to balance when the benefits of IE11 removal will materialize with the need to ensure ecosystem stability and support. If however the community feels strongly that we should move ahead on a more aggressive timeline, or if there are reasons to delay the actual removal further, we'd like to hear from you. Please provide the arguments and use cases in comments below.

This RFC will close on May 10, 2021.

# The benefits

Angular is an evergreen platform, meaning that it stays up to date with the evolving Web ecosystem. By removing support for legacy browsers we can focus our efforts on providing modern solutions and better support to developers and users.

There are several groups of stakeholders that will benefit from the removal of IE11 support:

- Application users
  - the apps will be smaller and load faster thanks to the removal of IE11-specific code
  - the use of modern Web APIs will result in further UX improvements

- Application authors and stakeholders
  - will benefit from improved APIs and build infrastructure
  - will reduce the risks and liabilities due to known security issues and limitations of IE11
  - can lean on our decision as an argument when talking to application stakeholders about IE11 support (if they still support it)

- Library authors
  - can lean on our decision and drop IE11 support in their libraries as well (if they still support it)

- Angular team
  - We'll be able to remove special JS and CSS code paths, polyfills, build passes, and dev infrastructure that exists only because we support IE11.
  - Since IE11 is the last non-evergreen browser we support, we'll gain access to many new modern web APIs and features that have been around for years, but we couldn't use them due to IE11 support and limitations to polyfilling. These APIs include:
    - CSS Variables, Intersection Observers, CSS Grid ([components/issues#7374](https://github.com/angular/components/issues/7374))
    - Native version of EcmaScript APIs (e.g. [regexp](https://github.com/angular/angular/blob/e7453f15d5b6925ff7be30993882f0c4a712eb71/packages/common/src/pipes/case_conversion_pipes.ts#L44-L54))
    - Web animation APIs
    - Proxies
    - etc

# Background information
## Microsoft support

IE11 was initially released almost 8 years ago, in 2013, and since that time, it has been receiving only bug and security fixes. It was superseded by Microsoft Edge, which unlike IE11 is an evergreen browser, which automatically receives feature updates including web platform API enhancements.

On August 17, 2020 [Microsoft announced deprecation of IE11 support in Microsoft 365](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666) apps and services, with the removal date set for August 17, 2021. Microsoft Teams already removed support for IE11 on November 30, 2020.

## Web API feature gap

The difference in Web APIs & features supported by IE11 compared to evergreen (modern) browsers is too vast to list in this RFC, and can instead be [reviewed at caniuse.org](https://caniuse.com/?compare=ie+11,edge+90,firefox+89,chrome+93,safari+14&compareCats=all).

## Browser usage

![StatCounter-browser-ww-monthly-202103-202103-bar](https://user-images.githubusercontent.com/216296/116285933-86bb4780-a743-11eb-9e9f-a2fc1075f6cb.png)

source: [statcounter.com](https://gs.statcounter.com/browser-market-share#monthly-202103-202103-bar)

<img width="1217" alt="Screen Shot 2021-04-26 at 10 32 44 AM" src="https://user-images.githubusercontent.com/216296/116285961-90dd4600-a743-11eb-9084-2ed8e29a275b.png">

source: [netmarketshare.com](https://netmarketshare.com/browser-market-share.aspx?options=%7B%22filter%22%3A%7B%7D%2C%22dateLabel%22%3A%22Custom%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22browserVersion%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22browsersDesktop%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222021-03%22%2C%22dateEnd%22%3A%222021-03%22%2C%22segments%22%3A%22-1000%22%7D)

**Comment by pavankjadda** [comment]
> The world-wide usage of Internet Explorer 11 (IE11) has been steadily declining for years, and is currently at 0.73% according to StatCounter and 1.70% according to NetMarketShare, meaning that very small fraction of the overall world population uses this browser and benefits from our effort to support it.

For this exact reason, we shipped first version of the project without IE 11 support. I would agree with you since Microsoft only supports Edge Chromium(for end users)

**Comment by penfold** [comment]
Really glad to see the proactive move away from IE11.

We registered with MS so that anyone visiting us with IE11 automatically redirect in to Edge.

https://docs.microsoft.com/en-us/microsoft-edge/web-platform/ie-to-microsoft-edge-redirection#request-an-update-to-the-ie-compatibility-list

**Comment by MaxArt2501** [comment]
I think IE usage should be deprecated as soon as Microsoft pulls the plug for its support for consumers, and removed soon after (next version, maybe?).
If someone needs to still support Internet Explorer, they could either provide the needed polyfills or use an older version of Angular.

Time to (finally) move on.

**Comment by e-oz** [comment]
Please consider this option: users who need IE11 support have an option to include tones of polyfills, but it will allow their apps to work.

Some companies require IE11 support because of “thin clients” they use. They don't care what will be the size of the application since app is for internal use. Leaving them with the outdated framework is not an option - it will not pass a security audit.

So with the “download tones of your polyfills” option, they would still have the ability to use Angular.

**Comment by paulparas** [comment]
Good decision, we spend most of our UI development time in fixing that one small control which works in chrome but not in IE & then we spend a lot of our precious time in applying hacks to make it work in IE. 

Clients want us to use latest angular version with IE support just in case someone in third world uses IE. Glad, now they have to chose either latest angular version or IE support. I think timeline is perfect and may even have been earlier as I don't know why someone would use IE in 2021 when they have chromium based edge or chrome. I have a top end i7 laptop and IE still fails to load fast. 

If its just because some people are used to IE and don't want to try anything else , then they need to be forced to use other browsers as otherwise they would keep using IE for another 100 years.

**Comment by IgorMinar** [comment]
> So with the “download tones of your polyfills” option, they would still have the ability to use Angular.

@e-oz there thank you for your feedback. Angular does all of this today and offers differential loading to minimize penalty for modern browsers. But even with this approach there is many web APIs we wish we could use but simply can't because they would be incompatible with IE11.

There are real limitations to what can be polyfilled in production-ready (secure, reliable, and performant) way. The same applies to transpilation. And for some APIs its better not to use them at all than to put in place a complex and slow build pipeline that would make things compatible.

**Comment by mhartington** [comment]
We have long since dropped IE11 support in Ionic and support this as well.

**Comment by yharaskrik** [comment]
I support this wholly and as quickly as possible. I realize there may be some uses cases that require ie11 but we have to move on and allow angular to evolve as a framework.

**Comment by mlc-mlapis** [comment]
Who says, `We need IE because ...` or `It's too expensive to migrate from IE to...` ignore the fact that those others pay all expenses (time, money, ...), and it has been going on for a long time, and it's not possible to continue by the same way.

**Comment by sgravrock** [comment]
Selfishly I'd like to see this go ahead, because it's one of the things that probably has to happen before Jasmine can drop IE support.

When my clients ask for IE support it's because they need browser upgrades in their organization to be infrequent and planned, and they haven't figured out where else to get that. Dropping IE might be more palatable to those organizations if more than one Firefox ESR version is supported.

**Comment by alisaduncan** [comment]
This feels like a reasonable and generous timeframe to me. 

The B2B app I work on has ~1.7% IE11 users we're working to message transitioning. Being able to point to Microsoft apps and significant dev frameworks like Angular dropping support will help lend credibility to the data we present in our transition messages.

I appreciate the Angular team asking for input from the community on these major decisions!

**Comment by jonparker** [comment]
I have previously worked with a company that built an internal enterprise app that only worked in an older version of Chrome. This meant they had to install the specific version of Chrome on employees machine's.
The reverse is also true. If IE11 compatibility is dropped then enterprises will come up with workarounds such as a side-by-side install of two browsers. There may be some work to modernize the integration of apps (e.g. provide a solution on how to redirect from an app that only supports IE to an app running in Edge/Chrome) but this is better than the extra work the community is doing to keep supporting IE11.

**Comment by andlewis** [comment]
Kill it with fire.

**Comment by lacolaco** [comment]
Although it is a bit anomalous, I think there is an option to extend the v12 LTS until the end of December 2022.
AngularJS supports IE11 and is maintained until the end of this year; the migration to Angular is expected to continue until the very last minute (or until early next year if they don't make it). We can also expect it to be difficult for them to upgrade directly from AngularJS to the latest Angular version.
I think that having v12 maintenance for at least 1 year after the end of AngularJS support is a deadline that can be balanced when considering the whole Angular ecosystem.

**Comment by IgorMinar** [comment]
> Kill it with fire.

There is no need for violence. 😊

**Comment by IgorMinar** [comment]
> I think that having v12 maintenance for at least 1 year after the end of AngularJS support is a deadline that can be balanced when considering the whole Angular ecosystem.

@lacolaco that would add more than a year if LTS to v12 which is a significant commitment, but we will consider it.

Can you help me understand the connection between AngularJS upgrades and IE11. In my experience they are two distinct problem spaces which sometimes overlap in some enterprises with a lot of legacy apps and clients.

If an enterprise has a fleet of IE11 clients, I'd hope that they would be working on updating the fleet in parallel to AngularJS upgrade. Upgrading to Angular is not meant to provide "indefinite" IE11 support.

Maybe I'm missing something?

**Comment by wilmarques** [comment]
I totally agree with this decision.

But I have a big problem.
I work at a bank and we have some apps that are used on some other companies.

Sometimes those companies don't have time/money/reason or even the desire to ditch IE, wo we just have to stick with it.
I agree that the number of users are declining, but sometimes a client is big enough to hold an entire business. 

I'll definitely use this as an argument:

> can lean on our decision as an argument when talking to application stakeholders about IE11 support (if they still support it)

It may end up well. But if not, we will have to move away from Angular on some specific apps.

Let's see what happens.

**Comment by lacolaco** [comment]
@IgorMinar My thought is, the LTS version which is living at the end of AngularJS should have the same IE11 support because LTS-to-LTS migration should be supported. 

> Angular v12 is planned to be the last version of Angular that supports IE11, this version will be:
> 
> released in May 2021
> actively supported through November 2021, and
> supported under LTS through November 2022

In the timeline, December 2022 is a gap. Then AngularJS is living but Angular doesn't support IE11. LTS-to-LTS migration won't be safe.

**Comment by IgorMinar** [comment]
> Let's see what happens.

@wileymarques please let us know how it goes. One reason for this RFC is to spark these kinds of conversations. Removing IE11 support is just a matter of time regardless of the framework/application, and starting the conversation now will avoid surprises in the future.

**Comment by wileymarques** [comment]
> Removing IE11 support is just a matter of time regardless of the framework/application.

@IgorMinar, indeed! I totally agree with you.

I'll try to get a response as soon as I can!

**Comment by adisreyaj** [comment]
I guess its the right time to do this. This has to be eventually done. +1 to what @IgorMinar said.

**Comment by pavankjadda** [comment]
> I totally agree with this decision.
> 
> But I have a big problem.
> I work at a bank and we have some apps that are used by some companies.
> 
> Sometimes those companies don't have time/money/reason or even the desire to ditch IE, so we just have to stick with it.
> I agree that the number of users are declining, but sometimes a client is big enough to hold an entire business.
> 
> I'll definitely use this as an argument:
> 
> > can lean on our decision as an argument when talking to application stakeholders about IE11 support (if they still support it)
> 
> It may end up well. But if not, we will have to move away from Angular on some specific apps.
> 
> Let's see what happens.

What would happen when Microsoft ditches IE 11 support?

**Comment by IgorMinar** [comment]
> What would happen when Microsoft ditches IE 11 support? 

They already have for MS Teams and will soon for MS 365 which represents most of their web products. Dropping support for the browser itself is likely complicated because of how the support for the browser is tied to the support for the OS and therefore won't formally happen for many more years. It's not reasonable for the web ecosystem to wait that long.

**Comment by SanderElias** [comment]
@IgorMinar It's about time. I dropped IE support for my customers when it was at version 8.  Mostly for 2 reasons.
1. extra development time (costs, this was usually enough to convince them)
2. security. IE is a known security risk.

@wileymarques I have had this discussion before. By now, the discussion is settled. Especially when you are a bank ;)
let me explain in 1 word. liability. 
Ok, let me extend on that a bit. A bank allowing their customers to use IE is telling their customers it is ok security-wise. I don't have to explain it is _not_ and at the time of this writing, this _is_ common knowledge. 
I'm pretty sure someone that got hacked (while using IE) will sue a bank pretty soon. I'm not a lawyer, but I'm pretty convinced they _do_ have an actual case.

It is about time IE gets dropped for the hot potato it is ;)

**Comment by naveedahmed1** [comment]
Great decision! The timeline seems quite reasonable. We recently dropped support for IE on Mustakbil.com.

After reading comment from @penfold , we've also registered our website with Microsoft's IE compatibility list, so that anyone visiting our website with IE11 is automatically redirected in to Edge.

**Comment by laiseng** [comment]
yes!!!!!!!!!!! finally, been waiting to erase my painful memory of IE11 hacks and workaround

**Comment by vijhhh2** [comment]
Please don't do this. That 1.7% might be employees from the company which I am working. Until now I some how manged and added support to ie11. If angular it's self remove the feature that means we cannot support ie11 that means I cannot upgrade whole bunch of apps to Angular 13 and our last update will be angular 12

**Comment by aaronczichon** [comment]
Dropping support for IE would be nice. The benefits of dropping support are way higher than keeping it in support.

**Comment by cyrilletuzi** [comment]
As an open source author, I support this move, as supporting IE is also a lot of work for lib authors (for example, one of my lib uses `indexedDb`, which has quite some bugs and missing features in IE).

This is especially annoying for open source authors working for free, while the very few enterprises left with IE are often companies with a lot of money.

I'd also like to link this RFC with the one about the end of Protractor: most of the newer e2e testing tools do not support testing in IE at all. So it means that while Angular CLI v12 will no longer install an e2e tool, I can't move to a newer tool yet in my lib as long as I have to support IE (where testing is the most important, because of the bugs / missing features).

**Comment by marco-bressi-deltatre** [comment]
For all companies/developers developing web frontends, supporting IE11 browser is not worth it nowadays.
Great decision!

**Comment by sbrez** [comment]
For all companies/developers developing web frontends, supporting IE11 browser is not worth it nowadays.
Great decision!

**Comment by willi84** [comment]
basically cool, but ...

* **data:** do we have data of how many applications still have to have IE support as requirement? 
* **migration guide:** would it be a good idea to give an overview of what will be thrown out for non-IE support or where IE is incompatible?
* **workaround:** maybe provide an alternative workaround like a plugin or using SSR/scully (lot of stuff probably just need the pre-rendered html and no extra application)
* **minimum app:** maybe have a basically IE compatible view running, who ask the user to use <modern browser> like the JS disabled mode

**more thoughts:**
* as far as I remember from my [bundle size project](https://github.com/willi84/ng-bundle-optimizer) is, that zone.js/webpack was the reason why stuff was or was not working in IE. 
* maybe we could provide a zone-less migration path with [rx-angular ](https://github.com/rx-angular/rx-angular)?

**Comment by mlc-mlapis** [comment]
@willi84 Hmm, a lot of work with minimal effect. It's not about zone.js/webpack. It's about the incompatibility of IE with modern browser technologies. All know the situation for many years, and they had a lot of space to prepare. The Angular team also has limited resources and time. If they should spend time on such things, what should others say?

**Comment by zbagley** [comment]
We decided to no longer offer full support for IE starting last year, shortly after Microsoft announced the push for Edge and the deprecation timeline of IE (including no more non-critical updates). Our target support specifically includes Chrome/Edge/Safari/Firefox.

**Comment by willi84** [comment]
@mlc-mlapis its just what I discovered while bringing an angular app to its minimal point.  even with this, at those time ivy demo setup. its not just about modern browser technologies. which is fun with the jam-stack getting back to good old HTML :-D 

but for sure it depends on your angular app complexity. 

and as I said: do we have data about it? I always have in mind  news articles from british government who used IE6 also years after it shoudl be already dead and for them it was - why ever - hard to update. 

compared to react Angular was meant for the business environment. and business environments often have harder upgrad pathes and use technologies for longer because of dependencies.

but even react provides polyfill plugins.  And they are using shadow dom:
https://inthetechpit.com/2020/05/11/make-react-app-compatible-with-ie11/

for me its hard to discuss if I dont know in detail which parts of angular's own code is incompatible with IE, why they need to be upgraded, how big is the outcome and so on. 

the stuff written in https://github.com/angular/components/issues/7374 is cool, but not sounds a lot like sugar on the top

**Comment by chaosmonster** [comment]
I like the time line of this RFC and support it.
Furthermore I would like to make an extending suggestion. As some companies can't get rid of IE11 for different reasons why not introduce another form of LTS.

The current LTS is provided by the Angular team. And after that time frame a community driven LTS could jump into place.

Basic idea of an LTS stays. No new features, no back porting of features only security updates and bugfixes. The difference here would be that it's fully community driven. Maybe even make this clear via the npm bundle download?

The Angular team provides the base for it. The caretaker will be a member of the community (in the beginning supported by the Angular Team). The MRs will be done by community members. 
This way the companies that still need to provide IE11 support can plan an exit strategy and support the open source community.

I know there is still a lot of stuff to be defined, discussed in this approach. But I feel like this could be an interesting compromise

**Comment by TPJ11** [comment]
This seem to be a classic chicken and the egg issue for those who are still using IE11 (almost all if not all are enterprise) they won't move off IE as "it still works" but when you talk about making it no longer work "you can't do that we need it!". 

To me at some point you have to remove support for IE11 and those who have yet to move off IE11 will not do it until they are forced to. Let's put it this why, no company is going to go "Angular no longer supports IE we are going to rewrite our application on a framework that does" you are not going to lose anyone from doing this and in the end its a needed step forward. 

If IE was a dog you would put it down.

**Comment by alsoicode** [comment]
Our web application is delivered through an EMR (Electronic Medical Records) application, which internally uses a web view that is an IE 11 window. The EMR vendor is moving away from that, and to a Chromium-based solution, but rollout won't be until at least 2023.

Losing IE 11 support will be unfortunate, but I agree it's way past time to move on.

**Comment by wileymarques** [comment]
> I'm pretty sure someone that got hacked (while using IE) will sue a bank pretty soon. I'm not a lawyer, but I'm pretty convinced they _do_ have an actual case.

@SanderElias, that's a fair point!

> - security. IE is a known security risk.

@IgorMinar, what do you think of documenting this? It may help some people trying to convince their bosses.

**Comment by naveedahmed1** [comment]
For those who still have to support IE, I believe its well supported in Angular 11.x, which is very stable and performant.
What if they keep their applications on V 11.x or even V 12.x which would have LTS through November 2022?

On a light note, I doubt that the clients who are so lazy to upgrade their browser, would be so demanding to have their apps upgraded to a newer version of Angular as soon as Angular releases new version. So, I believe it would be important to also consider the upgrade cycle of such apps, how often and quickly such apps are upgraded to new version of Angular.

**Comment by IgorMinar** [comment]
> > * security. IE is a known security risk.
> 
> @IgorMinar, what do you think of documenting this? It may help some people trying to convince their bosses.

@wileymarques I updated the benefits section for app developers with:

>   - will reduce the risks and liabilities due to known security issues and limitations of IE11

Is that sufficient?

**Comment by wileymarques** [comment]
@IgorMinar, I was thinking about something on https://angular.io/docs.
But I'm not sure if it's Angular responsibility.

**Comment by madebyjeffrey** [comment]
Although I am sure this has to happen eventually, a lot of our clients are not savvy users and are still running IE11. I would love to be free of IE11 support, but unfortunately that low global usage rate skyrockets when you narrow it down to certain groups of people. Given that large enterprises don't tend to move fast, it would be ideal to have some more time after Microsoft's own ceasing of support.

**Comment by mlc-mlapis** [comment]
@madebyjeffrey It will be a neverending story and who will pay the bill for others?

> they won't move off IE as "it still works"
> those who have yet to move off IE11 will not do it until they are forced to

**Comment by daniloff200** [comment]
I can't believe, that I will see that with my own eyes :eyes:

**Comment by MaxArt2501** [comment]
> Given that large enterprises don't tend to move fast, it would be ideal to have some more time after Microsoft's own ceasing of support.

@madebyjeffrey And how much time are we talking about here? Edge came out in 2015 - more than a year before Angular 2 was released - and it was clear it would have sunset IE. Enterprises should have started transitioning long ago.

I recently had to made an app for an enterprise that forced me to use their own framework that's compatible with Angular 7 only, so I know the drill quite well. They have "plans" to update it to Angular *10*, but I'm not holding my breath.

The point is: those kind of enterprises don't care about the latest version of a front-end framework and won't mind being stuck with Angular 11 for a while. It's usually developers that do.

The only thing thats that could make a company move are money loss and security issues (that could lead to money loss, or legal troubles and eventually money loss). And well, it's much more plausible that IE will raise security concerns in the next months rather than an extended use of Angular 11.

**Comment by eemt** [comment]
It's a moral imperative to drop support for IE 11.  Supporting IE 11 is supporting [security risks and unsatisfactory user experiences](https://thinkadnet.com/2020/10/top-5-reasons-to-stop-using-internet-explorer/).  [Microsoft themselves have made it clear](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/the-perils-of-using-internet-explorer-as-your-default-browser/ba-p/331732):  Customers that need IE 11 for legacy applications, should use it _only_ for those.  Do not use it for any other reason.  It must not be used as a default browser.  All applications should stop supporting IE 11 [just as Microsoft is doing](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).

**Comment by ocombe** [comment]
finally :) honestly people who still need IE11 should just not upgrade Angular to a new version and that's all, they shouldn't block everyone else from getting a better framework
I would even go so far as to guess that it might already be the case because a lot of libraries have dropped IE11 support, and they can't upgrade to new version of those libraries, which might block them from updating Angular anyway.

**Comment by IgorMinar** [comment]
@ocombe thanks for weighing in. From the discussion with various companies, we know that there are challenges around using unsupported versions of software in some industries, especially in government and healthcare. In order to support the developers in these industries we prefer the proposed deprecation & removal timeline. We know that many of these companies would prefer indefinite support, but that's just not feasible, so we are hoping to find a good compromise between removing IE11 support immediately and many years in the future.

**Comment by ocombe** [comment]
@IgorMinar I know that from a PR perspective it's always better if you can support more browsers, that being said it's also important when frameworks push the web forward by dropping old browsers. Fyi Vue is also dropping IE11 support: https://github.com/vuejs/rfcs/discussions/296#discussioncomment-669877

**Comment by BuZZ-dEE** [comment]
die IE11

**Comment by wileymarques** [comment]
For some companies, IE removal is definitely going to be a hassle. If you see my comments above, I try to say that.

But, I don't think those companies should hold an entire industry. The web needs to move forward, and so Angular!

**Comment by chaosmonster** [comment]
@IgorMinar can you by any chance tell if those companies, would be fine with a **community driven LTS** after November 2022?

**Comment by IgorMinar** [comment]
> @IgorMinar can you by any chance tell if those companies, would be fine with a community driven LTS after November 2022?

@chaosmonster I don't believe so. I think they would be more inclined to consultancy-backed paid LTS support. But if anyone from banking, government, or healthcare industries has a different opinion, it would be great if they shared it here.

**Comment by d-koppenhagen** [comment]
Once the IE support is removed, will there be a switch in the `ListKeyManager` implementation from `event.keyCode` to `event.code` of the CDK? As currently IDEs will provide a deprecated issue like this:

<img width="963" alt="Screenshot 2021-04-29 at 18 11 46" src="https://user-images.githubusercontent.com/4279702/116583209-65f71b80-a916-11eb-8193-f09c49e67e0e.png">

See also closed issue here: https://github.com/angular/components/issues/12672

**Comment by itea-dev** [comment]
Why not remove official support for IE early in v12, instead of waiting to v13?

Companies interested in IE can stay in version 11 which still has more than 1 year of support.

**Comment by kaito3desuyo** [comment]
I hate IE as much as you do, and I agree with the community policy of removing IE support.

Unfortunately, the usage rate of IE is still high in Japanese public organizations (about 30-40% by experience), and it is used within the intranet, so security reasons do not motivate the renewal of devices and browsers. Some organizations even have a security policy of "using IE11".

And even more unfortunately, I and my team are developing an app in Angular for use on that intranet...

I think the immediate solution here is to "do not update from Angular v11".
It would be nice to have the ability to optionally specify the version instead of always installing the latest version (corresponding to the Angular CLI version) with the "ng update" command.

**Comment by indraraj26** [comment]
I would suggest create one additional package, @angular/ie-11-support

it should load before bootstrap (main.ts)

there are banks/upi solution which still uses IE11, it is really painful to explain them.

**Comment by IgorMinar** [comment]
> I would suggest create one additional package, @angular/ie-11-support

@indraraj26 this would essentially mean polyfilling the missing APIs, which is what we do today but that only works for a subset of APIs we'd like to use. For example it's not possible to polyfill CSS custom properties in this way.

**Comment by IgorMinar** [comment]
> It would be nice to have the ability to optionally specify the version instead of always installing the latest version (corresponding to the Angular CLI version) with the "ng update" command.

@kaito3desuyo you can use this trick: `npx @angular/cli@11 new my-angular-v11-app`

**Comment by spock123** [comment]
We specifically do not support any IE version at all. We never had any issues with that - removing support greatly simplifies and improves our solution.

**Comment by waynon** [comment]
I think the timeline of dropping IE 11 support can be more aggressive since those who still need IE 11 support can stay on Angular 11 as it's stable and performant enough and will be supported in almost 1 year. Considering the benefits of dropping IE 11 support, we can drop IE 11 support in Angular 12 immediately instead of waiting for Angular 13.

**Comment by waynon** [comment]
I think the timeline of dropping IE 11 support can be more aggressive since those who still need IE 11 support can stay on Angular 11 as it's stable and performant enough and will be supported in almost 1 year. 

Considering the benefits of dropping IE 11 support, we can drop IE 11 support in Angular 12 immediately instead of waiting for Angular 13.

**Comment by IgorMinar** [comment]
Just a quick summary of the discussion and a reminder that the RFC closes next Monday.

So far it seems that vast majority of the people that got involved in the discussion support the deprecation and removal timeline as proposed. There are two additional smaller groups of people — those that would like us to outright remove IE11 support in v12, and those that would like to see an extension of the support period.

Several commenters asked if we could offer polyfills and additional transpilation — this is something we already heavily do and that's how the current IE11 support works, but there are real limitations to what we can do with this approach and we feel strongly that these limitations are holding the entire Angular ecosystem back and prevent all of us taking advantage of using modern Web APIs.

In https://github.com/angular/angular/issues/41840#issuecomment-828111512 @SanderElias reminded us that there are real security threats and risks affecting any businesses that still rely on IE11 and these concerns outweigh concerns about teams eventually using unsupported version of Angular.

If you have additional comments or things that we missed, please let us know. Thank you!

**Comment by Toledochrist** [comment]
Thanks

**Comment by s-gbz** [comment]
I am really glad the Angular team is taking a step into a modern, IE free web :rocket: :fire: !
I also do agree with suggestions as @waynon s (tackle the process even faster) since the only users of IE are "old fashioned" companies that rarely update their systems anyway (my humble experience). They _have_ found workarounds for IE and probably will find further if necessary.

This is probably not the most objective or persuasive argument, but HEY, we all know there's truth in it :smile: :innocent:

**Comment by eemt** [comment]
One follow up comment.  For those who believe they "need IE support for legacy applications."  Microsoft has provided a path forward: Edge/legacy mode.  If that doesn't work, that's Microsoft's responsibility to address.  It shouldn't hold up everyone else.

**Comment by 9211rishu** [comment]
I totally support the move that is being taken up and timeline also seems perfect. I don't know who is using IE in 2021. Great decision.

**Comment by IgorMinar** [comment]
This RFC is now closed, thank you everyone for participating.

Since my [last summary of the discussion](https://github.com/angular/angular/issues/41840#issuecomment-832154744) we heard from several developers that they'd like IE11 support removed in v12. Judging by the number of reactions, it seems that this group of community members is still in minority compared to those that would prefer to go with the originally proposed timeline.

In reality the difference between dropping IE11 in v12 and v13 at this point in time is negligible. The master/main branch will be open for breaking changes targeting v13 in 2-3 months, so that's when we we'll able to start making IE11 incompatible changes. 2-3 months is insignificant in the long term perspective, and not worth creating the extra inconveniences for developers and teams that still support IE11.

So to summarize, the originally proposed timeline is the one with the most support, this means that IE11 support will be deprecated in v Angular v12 (to be released in May 2021 and supported through November 2022), and IE11 support removal will occur in Angular v13 (late 2021).

Thank you again everyone for sharing your thoughts with us and helping us make the decision.

**Comment by Stargator** [comment]
@IgorMinar thanks for opening this. I wished I had seen it before the ng 12 release, but I'm happy that the team engaged in this discussion.

**Comment by pavankjadda** [comment]
https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/

**Comment by michaelrobertson-fico-com** [comment]
Fully support this decision. Frankly it's difficult to understand the tenacity with which some organizations have clung to this relic. There have been much better options for a very long time. Our stack includes Bootstrap and we were supportive of BS5's decision to drop IE11 support. It's great to have clarity from our framework of choice.
